Support KAS

kkimurak opened the pull request:
This PR adds configurations for external / built-in KAS and installation process to build built-in KAS.

Close #2736 

External KAS settings implemented by @antt1995 (in #2598, rebased by @kkimurak)
Built-in KAS build process implemented by @kkimurak  
Built-in KAS configuration test, improvements and updates until it works by @th-2021 

TODO

- [x] Documentation cleanup, verify that the actual behavior matches the documentation

antt1995 committed:
KAS

antt1995 committed:
KAS

antt1995 committed:
Added functions

antt1995 committed:
Added functions

antt1995 committed:
Update assets/runtime/env-defaults

Co-authored-by: Kazunori Kimura <33391846+kkimurak@users.noreply.github.com>

antt1995 committed:
Update assets/runtime/env-defaults

Co-authored-by: Kazunori Kimura <33391846+kkimurak@users.noreply.github.com>

antt1995 committed:
Added to Readme

antt1995 committed:
Added to Readme

antt1995 committed:
Update assets/runtime/env-defaults

Co-authored-by: Kazunori Kimura <33391846+kkimurak@users.noreply.github.com>

antt1995 committed:
Update assets/runtime/env-defaults

Co-authored-by: Kazunori Kimura <33391846+kkimurak@users.noreply.github.com>

antt1995 committed:
Update README.md

antt1995 committed:
Update README.md

antt1995 committed:
Update README.md

Co-authored-by: Kazunori Kimura <33391846+kkimurak@users.noreply.github.com>

antt1995 committed:
Update README.md

Co-authored-by: Kazunori Kimura <33391846+kkimurak@users.noreply.github.com>

kkimurak committed:
README.md: revise GitLab KAS section

Fix link fragment for gitlab-kas in TOC
(markdownlint: MD-051 - must be lowercase)

kkimurak committed:
README.md: revise GitLab KAS section

Fix link fragment for gitlab-kas in TOC
(markdownlint: MD-051 - must be lowercase)

kkimurak committed:
Add built-in KAS

- build KAS on build, enable via GITLAB_AGENT_KAS_ENABLED
  (automatically enabled if GITLAB_KAS_ENABLED=true)
- Add built-in KAS config file
  kas config: import upstream (v15.10.0 - d88f4b89)
- minimal parameterization
  gitlab side:
  - GITLAB_KAS_ENABLED
    gitlab_rails['gitlab_kas_enabled'] for omnibus installation
  - GITLAB_KAS_INTERNAL
    gitlab_rails['gitlab_kas_internal_url'] for omnibus installation
  - GITLAB_KAS_EXTERNAL
    gitlab_rails['gitlab_kas_external_url'] for omnibus installation
  - GITLAB_KAS_PROXY
    gitlab_rails['gitlab_kas_external_k8s_proxy_url'] for omnibus installation
  kas side:
  - GITLAB_AGENT_KAS_ENABLED
    gitlab_kas['enabled'] for omnibus installation
  - GITLAB_AGENT_KAS_API_LISTEN_AUTHENTICATION_SECRET_FILE
    gitlab_kas['internal_api_listen_authentication_secret_file']
  - GITLAB_AGENT_KAS_PRIVATE_API_LISTEN_AUTHENTICATION_SECRET_FILE
  used by both:
  - GITLAB_KAS_SECRET
    used as a value for
    - gitlabhq/gitlab.yml : production.gitlab_kas.secret_file
    - gitlab-agent/gitlab-kas_config.yaml : gitlab.authentication_secret_file

Prefix for parameters that only used by gitlab-agent/kas is "GITLAB_AGENT_KAS_"

Also add process to generate secret files for KAS

Update built-in KAS to 16.10.1, sync config

Update built-in gitlab-kas to v17.0.2, sync configuration

- Bump to v17.0.0, sync configuration
  GitOps module have been removed. See upstream change:
  https://gitlab.com/gitlab-org/cluster-integration/gitlab-agent/-/merge_requests/1436

- Bump: gitlab-kas 17.0.1
  no change in configuration file

- Bump: gitlab-kas 17.0.2
  no change in configuration file

Personal note:
I am still unable to successfully connect the agent from the WebUI.
The problem is that I don't know the correct way to do it in the first place,
so I have to check it out.

Stop gitlab_kas before restoring backup

It seems that kas has DB connection and blocks restoration

- Bump: gitlab-kas 18.1.0
  reference configuration file have been renamed to kascfg_defaults.yaml
  and does not contain full configurations
  So I have to search a way to generate full configuration from .proto
  or docs

kkimurak committed:
Add KAS config file, minimal parameterization

also add process to generate secret files for KAS
- GITLAB_KAS_SECRET
- GITLAB_KAS_API_LISTEN_AUTHENTICATION_SECRET_FILE
- GITLAB_KAS_PRIVATE_API_LISTEN_AUTHENTICATION_SECRET_FILE

kkimurak committed:
Add KAS config file, minimal parameterization

also add process to generate secret files for KAS
- GITLAB_KAS_SECRET
- GITLAB_KAS_API_LISTEN_AUTHENTICATION_SECRET_FILE
- GITLAB_KAS_PRIVATE_API_LISTEN_AUTHENTICATION_SECRET_FILE

kkimurak committed:
Add built-in KAS

- build KAS on build, enable via GITLAB_AGENT_KAS_ENABLED
  (automatically enabled if GITLAB_KAS_ENABLED=true)
- Add built-in KAS config file
  kas config: import upstream (v15.10.0 - d88f4b89)
- minimal parameterization
  gitlab side:
  - GITLAB_KAS_ENABLED
    gitlab_rails['gitlab_kas_enabled'] for omnibus installation
  - GITLAB_KAS_INTERNAL
    gitlab_rails['gitlab_kas_internal_url'] for omnibus installation
  - GITLAB_KAS_EXTERNAL
    gitlab_rails['gitlab_kas_external_url'] for omnibus installation
  - GITLAB_KAS_PROXY
    gitlab_rails['gitlab_kas_external_k8s_proxy_url'] for omnibus installation
  kas side:
  - GITLAB_AGENT_KAS_ENABLED
    gitlab_kas['enabled'] for omnibus installation
  - GITLAB_AGENT_KAS_API_LISTEN_AUTHENTICATION_SECRET_FILE
    gitlab_kas['internal_api_listen_authentication_secret_file']
  - GITLAB_AGENT_KAS_PRIVATE_API_LISTEN_AUTHENTICATION_SECRET_FILE
  used by both:
  - GITLAB_KAS_SECRET
    used as a value for
    - gitlabhq/gitlab.yml : production.gitlab_kas.secret_file
    - gitlab-agent/gitlab-kas_config.yaml : gitlab.authentication_secret_file

Prefix for parameters that only used by gitlab-agent/kas is "GITLAB_AGENT_KAS_"

Also add process to generate secret files for KAS

Update built-in KAS to 16.10.1, sync config

Update built-in gitlab-kas to v17.0.2, sync configuration

- Bump to v17.0.0, sync configuration
  GitOps module have been removed. See upstream change:
  https://gitlab.com/gitlab-org/cluster-integration/gitlab-agent/-/merge_requests/1436

- Bump: gitlab-kas 17.0.1
  no change in configuration file

- Bump: gitlab-kas 17.0.2
  no change in configuration file

Personal note:
I am still unable to successfully connect the agent from the WebUI.
The problem is that I don't know the correct way to do it in the first place,
so I have to check it out.

Stop gitlab_kas before restoring backup

It seems that kas has DB connection and blocks restoration

- Bump: gitlab-kas 18.1.0
  reference configuration file have been renamed to kascfg_defaults.yaml
  and does not contain full configurations
  So I have to search a way to generate full configuration from .proto
  or docs

kkimurak committed:
kas: tweak defaults for gitlab.authentication_secret_file

default to api.listen.authentication_secret_file,
as omnibus-gitlab do

kkimurak committed:
kas: tweak defaults for gitlab.authentication_secret_file

default to api.listen.authentication_secret_file,
as omnibus-gitlab do

kkimurak committed:
Tweak kas config order in gitlab.yml

sync with upstream

kkimurak committed:
Tweak kas config order in gitlab.yml

sync with upstream

th-2021 committed:
upgrade to 18.5.1

squash following commit:
- update to 18.5.0
- upgrade to 18.5.1

th-2021 committed:
upgrade to 18.5.1

squash following commit:
- update to 18.5.0
- upgrade to 18.5.1

th-2021 committed:
change name of secret

add WEBTOKEN secret; remove GITLAB_KAS_SECRET

Replace GITLAB_KAS_SECRET by GITLAB_AGENT_KAS_WEBSOCKET_TOKEN_SECRET_FILE

th-2021 committed:
change name of secret

add WEBTOKEN secret; remove GITLAB_KAS_SECRET

Replace GITLAB_KAS_SECRET by GITLAB_AGENT_KAS_WEBSOCKET_TOKEN_SECRET_FILE

th-2021 committed:
configure NGINX for GITLAB_BUILTIN_KAS

Add kas location to assets/runtime/config/nginx/gitlab(-ssl) config

KAS location is affected by GITLAB_RELATIVE_URL_ROOT
so add update process to assets/runtime/functions

th-2021 committed:
configure NGINX for GITLAB_BUILTIN_KAS

Add kas location to assets/runtime/config/nginx/gitlab(-ssl) config

KAS location is affected by GITLAB_RELATIVE_URL_ROOT
so add update process to assets/runtime/functions

th-2021 committed:
tweak agent.kubernetes_api.url_path_prefix

in gitlab-kas_config.yaml

complete 18.5.0 upgrade

th-2021 committed:
tweak agent.kubernetes_api.url_path_prefix

in gitlab-kas_config.yaml

complete 18.5.0 upgrade

th-2021 committed:
fix GITLAB_AGENT_BUILTIN_KAS_ENABLED == false

th-2021 committed:
fix GITLAB_AGENT_BUILTIN_KAS_ENABLED == false

kkimurak committed:
Fix parameter name to enable builtin-kas

Merge GITLAB_AGENT_BUILTIN_KAS_ENABLED and GITLAB_AGENT_KAS_ENABLED

kkimurak committed:
Fix parameter name to enable builtin-kas

Merge GITLAB_AGENT_BUILTIN_KAS_ENABLED and GITLAB_AGENT_KAS_ENABLED

th-2021 commented:
Should we make the builtin kas the only option? Remove GITLAB_AGENT_BUILTIN_KAS_ENABLED?

th-2021 committed:
fix kubernetes access

th-2021 committed:
fix kubernetes access

kkimurak commented:
@th-2021 Thanks, I cherry-picked 4a4c484 .  

For `GITLAB_AGENT_BUILTIN_KAS_ENABLED` removal, I'm not sure. Some people may want to continue using the existing KAS and connect it to `sameersbn/gitlab` (although I can't imagine many situations where KAS is running alone).

kkimurak committed:
Add description for `GITLAB_AGENT_KAS_WEBSOCKET_TOKEN_SECRET_FILE`

kkimurak committed:
Add description for `GITLAB_AGENT_KAS_WEBSOCKET_TOKEN_SECRET_FILE`

kkimurak committed:
Fix assignment for GITLAB_KAS_SECRET

User still only required to set
`GITLAB_AGENT_KAS_API_LISTEN_AUTHENTICATION_SECRET_FILE`
as `GITLAB_KAS_SECRET` finally default to
`GITLAB_AGENT_KAS_API_LISTEN_AUTHENTICATION_SECRET_FILE`
(over `GITLAB_AGENT_KAS_GITLAB_AUTHENTICATION_SECRET_FILE`)

kkimurak committed:
Fix assignment for GITLAB_KAS_SECRET

User still only required to set
`GITLAB_AGENT_KAS_API_LISTEN_AUTHENTICATION_SECRET_FILE`
as `GITLAB_KAS_SECRET` finally default to
`GITLAB_AGENT_KAS_API_LISTEN_AUTHENTICATION_SECRET_FILE`
(over `GITLAB_AGENT_KAS_GITLAB_AUTHENTICATION_SECRET_FILE`)

kkimurak committed:
Add example KAS configuration to README.md

kkimurak committed:
Add example KAS configuration to README.md

kkimurak commented:
Some documentation updated. I think it's ready for review.  

For those who interested in but cannot build themselves, I pushed `kkimurak/sameersbn-gitlab:18.7.0-kas` for testing. Any feedback would be appreciated.

th-2021 commented:
@kkimurak it's not working for me. The reason seems to be that gitlab cannot be reached at port 8080. Changing the port to 8181 in gitlab kas config seems to solve it. At least for kas itself. The gitlab-agent (k8s part) still cannot connect.

kkimurak commented:
Built successfully. pushed `kkimurak/sameersbn-gitlab:18.7.0-kas-e2673eb7`

kkimurak commented:
Rebased onto v18.8.2. Updated built-in KAS to 18.8.2 as well.  
Test image available: `kkimurak/sameersbn-gitlab:18.8.2-kas-85233586`

th-2021 commented:
Works for me (2 instances)

kkimurak commented:
Rebased onto v18.9.2. Updated built-in KAS to 18.9.2 as well. No update to the configuration file from v18.9.0.  
Test image available: kkimurak/sameersbn-gitlab:18.9.2-kas-64cfbfc9

kkimurak committed:
Sync gitlab-kas config to 18.5.1

- Don't mention IPv4 for localhost
  since it'll listen on both v4 and v6 anyway
  https://gitlab.com/gitlab-org/cluster-integration/gitlab-agent/-/merge_requests/2641
  Since: v18.2.0
- Improve kas and agentk configuration docs
  https://gitlab.com/gitlab-org/cluster-integration/gitlab-agent/-/merge_requests/2902
  (This sets default value for agent.listen.websocket to false,
  but this commit just ignored it)
  Since: v18.4.0

Also, we set non-default value (debug) for
observability.logging.level (default: info)
and .grpc_level (default: error)

kkimurak committed:
Sync gitlab-kas config to 18.5.1

- Don't mention IPv4 for localhost
  since it'll listen on both v4 and v6 anyway
  https://gitlab.com/gitlab-org/cluster-integration/gitlab-agent/-/merge_requests/2641
  Since: v18.2.0
- Improve kas and agentk configuration docs
  https://gitlab.com/gitlab-org/cluster-integration/gitlab-agent/-/merge_requests/2902
  (This sets default value for agent.listen.websocket to false,
  but this commit just ignored it)
  Since: v18.4.0

Also, we set non-default value (debug) for
observability.logging.level (default: info)
and .grpc_level (default: error)

kkimurak committed:
Update built-in KAS to 18.11.2

Just bump the version
unless configuration file pkg/kascfg/kascfg_default.yml is not updated

- Update to 18.8.2
- Update to 18.8.4
- Update to 18.9.0
- Update to 18.9.1
- Update to 18.9.2
- Update to 18.10.0
- Update to 18.10.1
- Update to 18.10.3
- Update to 18.11.0
- Update to 18.11.1
- Update to 18.11.2

kkimurak commented:
Rebased to v18.10.1. Updated built-in KAS to v18.10.1 as well.  
Test image available: `kkimurak/sameersbn-gitlab:18.10.1-kas-f1e8dcb8`

kkimurak commented:
Rebased to v18.10.3. Updated built-in KAS to v18.10.3 as well.  
Test image available: `kkimurak/sameersbn-gitlab:18.10.3-kas-eeb950fc`

kkimurak commented:
Rebased to v18.11.0. Updated built-in KAS to v18.11.0 as well.
Test image available: kkimurak/sameersbn-gitlab:18.11.0-kas-cfa47f80

th-2021 commented:
@kkimurak first time I tried to delete a pod using the environment I came across a CORS issue. KAS doesn't know the external gitlab url. The following patch solves it:
```
--- a/assets/runtime/config/gitlab-agent/gitlab-kas_config.yaml
+++ b/assets/runtime/config/gitlab-agent/gitlab-kas_config.yaml
@@ -51,6 +51,7 @@ gitaly:
     refill_rate_per_second: 15
 gitlab:
   address: http://127.0.0.1:8080{{GITLAB_RELATIVE_URL_ROOT}}
+  external_url: {{GITLAB_SCHEME}}://{{GITLAB_HOST}}
   authentication_secret_file: {{GITLAB_AGENT_KAS_GITLAB_AUTHENTICATION_SECRET_FILE}} # required
   api_rate_limit:
     bucket_size: 250
diff --git a/assets/runtime/functions b/assets/runtime/functions
index e2326823..3b58f6da 100644
--- a/assets/runtime/functions
+++ b/assets/runtime/functions
@@ -375,8 +385,16 @@ gitlab_configure_gitlab_kas() {
     GITLAB_KAS_PROXY

   printf "Configuring gitlab-agent::KAS (enabled: %s)\n" "${GITLAB_AGENT_BUILTIN_KAS_ENABLED}"
+
   if [[ ${GITLAB_AGENT_BUILTIN_KAS_ENABLED} == true ]]; then
+    if [[ "${GITLAB_HTTPS}" == true ]]; then
+      GITLAB_SCHEME=https
+    else
+      GITLAB_SCHEME=http
+    fi
     update_template ${GITLAB_KAS_CONFIG} \
+      GITLAB_SCHEME \
+      GITLAB_HOST \
       GITLAB_RELATIVE_URL_ROOT \
       GITLAB_AGENT_KAS_API_LISTEN_AUTHENTICATION_SECRET_FILE \
       GITLAB_AGENT_KAS_PRIVATE_API_LISTEN_AUTHENTICATION_SECRET_FILE \
```

th-2021 committed:
KAS: fix CORS issue by setting gitlab.external_url

th-2021 committed:
KAS: fix CORS issue by setting gitlab.external_url

kkimurak commented:
Rebased to v18.11.2. Updated built-in KAS to v18.11.2 as well.  
Included suggestion from @th-2021 as commit 94bfe01. Thanks.

Test image available: kkimurak/sameersbn-gitlab:18.11.2-kas-94bfe01c

kkimurak committed:
Update built-in KAS to 19.2.4

Just bump the version
unless configuration file pkg/kascfg/kascfg_default.yml is not updated

- Update to 18.8.2
- Update to 18.8.4
- Update to 18.9.0
- Update to 18.9.1
- Update to 18.9.2
- Update to 18.10.0
- Update to 18.10.1
- Update to 18.10.3
- Update to 18.11.0
- Update to 18.11.1
- Update to 18.11.2
- Update to 18.11.3
- Update to 19.0.1
- Update to 19.0.2
- Update to 19.1.0
- Update to 19.1.1
- Update to 19.2.0
- Update to 19.2.1
- Update to 19.2.2
- Update to 19.2.4

kkimurak commented:
Rebased on v19.2.2. Updated built-in KAS to v19.2.2 as well.
Test image available: [kkimurak/sameersbn-gitlab:19.2.2-kas-d0a0d046](https://hub.docker.com/repository/docker/kkimurak/sameersbn-gitlab/tags/19.2.2-kas-d0a0d046/sha256-8ff8ca6fb6f01b86efe6c92d9430d09c4863fcba22f0a819978db6aedd4eda3b)

kkimurak commented:
Rebased on v19.2.4. Updated built-in KAS to v19.2.4 as well.
Test image available: [kkimurak/sameersbn-gitlab:19.2.4-kas-ea9d2cb5](https://hub.docker.com/layers/kkimurak/sameersbn-gitlab/19.2.4-kas-ea9d2cb5/images/sha256-f3837addcccd4e0a80069ab1169f8c0ac19b1d78f6a48a630e42e0452058eca7)

kkimurak commented:
Rebased on v19.3.0, then v19.3.1. Updated built-in KAS to v19.3.1 as well.
Test image available:
- [kkimurak/sameersbn-gitlab:19.3.0-kas-ddd75e7a](https://hub.docker.com/layers/kkimurak/sameersbn-gitlab/19.3.0-kas-ddd75e7a/images/sha256-9ccacffd4d6868207ec45888bcf690e33081933a61e38da688ca2048bbc90bf7)
- [kkimurak/sameersbn-gitlab:19.3.1-kas-d8827b98](https://hub.docker.com/layers/kkimurak/sameersbn-gitlab/19.3.1-kas-d8827b98/images/sha256-40bc3e8fb80a50de84f65d100891255dc7b79986d57d53b9f83a2c18a07d0c2d)
